submariner/multiple-active-gateways.md

```diff
@@ -0,0 +1,100 @@
+# Multiple Active Gateways
+
+## Summary
+
+NOTE: The design details are still a work in progress, so this is currently a draft proposal.
+
+Submariner currently only allows a single gateway to be active at any one time in a given cluster.
+This enhancement proposes that there be an option to allow more than one gateway to be active in
+a given cluster to enhance both performance and resiliency. This feature is sometimes referred to as
+“active-active gateways”; however, the goal is to allow two or more actove gateways to be used.
+
+## Proposal
+
+The proposal is to optionally allow multiple gateways to be active at the same time in a given cluster.
+
+When only a single active gateway is used, all inter-cluster traffic must flow through that one gateway,
+which creates a performance bottleneck. Enabling more than one node to be used as a gateway can improve
+the performance linearly in the number of gateways used.
+
+Additionally, when the active gateway fails in the current implementation, work must be done to switch
+over to the new gateway including the establishment of new tunnels (both intra-cluster and inter-cluster)
+and new routes, which takes time. With the proposed design, It is expected that the fail-over times can
+be significantly improved if additional gateways are operational. Furthermore, it is expected that
+traffic flowing through operational gateways will continue to flow while the traffic from the failed
+gateway is redirected to the operational gateways. Both properties reduce the impact of a failure.
+
+## Design Details
+
+### Single Gateway
+
+In the current single gateway Submariner implementation, intra-cluster VXLAN tunnels are established by
+the non-gateway nodes to the gateway node, and each gateway establishes an inter-cluster tunnel to each
+other gateway in the clusterset (with the tunnel type determined by the cable driver in use) as shown
+below.
+
+![SingleGateway](./images/single-gateway.png)
+
+### Multiple Active Gateway Design
+
+The proposal for supporting multiple active gateway nodes is shown in the following diagram and described below.
+
+![MultiGateway](./images/multi-gateway.png)
+
+* Each non-gateway node will establish a VXLAN tunnel to each gateway node.
+* The non-gateway nodes will load-balance flows between the active gateway nodes.
+* Each gateway node will establish a tunnel to each peer gateway node.
+* The gateway nodes will load-balance flows between the active peer gateway nodes.
+* Load balancing will be performed using Linux resilient nexthop groups.
```

Comment:
Currently the Active Submariner Gateway pod is responsible for couple of things like
1. Creating local endpoint and cluster objects 
2. Syncing the endpoints/cluster objects with the Broker Cluster
3. Healthchecks with remote cluster gateway nodes
4. Performing NAT discovery to identify if remote cluster is reachable over private or public-ip in the endpoint.
5. Gateway object syncer.
6. Supporting metrics

Since some of the above operations have to be done by only one component, can you please add notes on how the above operations will be handled in the new Active/Active architecture.

Reply:
This has been described in the updated proposal.